GraphQL introduction (#655)
vegaprotocol/documentation: docs/api/graphql/overview.md

---
title: Using GraphQL
hide_title: false
sidebar_position: 1
---
GraphQL provides read-only queries and subscriptions for getting data from Vega. Support is built into the data nodes, and some Vega data nodes may have a GraphQL interface enabled. GraphQL is an alternative to REST that provides more flexibility for building custom queries.

> While REST provides multiple endpoints with small amounts of data, GraphQL provides a single endpoint that inputs complex queries and outputs as much information as is needed for the query.

Reviewer: ```suggestion
While REST provides multiple endpoints with small amounts of data, GraphQL provides a single endpoint that allows users to input complex queries and outputs as much information as is needed for the query.
```

  }
}
```

This would return something that looks like:
```
{
  "data": {
    "proposalsConnection": {
      "pageInfo": {
        "hasNextPage": true,
        "endCursor": "eyJzdGF0ZSI6NiwidmVnYV90aW1lIjoiMjAyMy0wNS0yNFQxNDowMDowMS4wMjk3MTZaIiwiaWQiOiI4NDAyNWU2ODM4N2NmNjFjMmI5MTIyOGQ3NjhkY2RkNGYxMGE5ZWU1Y2QyODI0ZmRlYTM1YjI1OTk3NmY1OWMxIn0="
      },
      "edges": [
        {
          "node": {
            "id": "84025e68387cf61c2b91228d768dcdd4f10a9ee5cd2824fdea35b259976f59c1",
            "rationale": {
              "title": "VMP-007 - Create market - LINK/USDT Future - 2023/06/30"
            }
          }
        }
      ]
    }
  }
}
```

There’s some added complexity here. Let’s break it down:

On the simple query, the structure of the result is basically identical to the query - so the shape is almost exactly what you've requested and nothing else.

Whereas with the paginated connections, the results contain extra nesting that isn't in the query. 

This nesting includes a `pageInfo` property with some information about the data set. Here, you can see that with the current limit of 1, there is another page. You can also see the cursor for the last item in the result.

> There's also an edges property that contains nodes. The data under node shows a proposal, information you would get by querying for a proposal by ID.

Reviewer: ```suggestion
There's also an edges property that contains an array of edges. Each edge is made up of a node and a cursor. The data under node shows a proposal, information you would get by querying for a proposal by ID, while the cursor is a unique pointer that identifies the node of data.
```

> Paginated results use this added nesting to provide the extra information you need to navigate the data. Each edge item in the array also has a cursor property you could select if you need to.

Reviewer: ```suggestion
```

> After fetching this result, it's clear that there are more proposals. You can use `endCursor` from the first result to fetch the next one, which will return the next 1 proposal, and a new cursor:

Reviewer: ```suggestion
After fetching this result, it's clear from the `pageInfo` that there are more proposals. You can use `endCursor` from the first result's `pageInfo`, or the cursor from the edge record (they should be the same) to fetch the next one, which will return the next 1 proposal, and a new cursor:
```

    midPrice
  }
}
```

### Bus event subscription 
The bus event subscription is different from other subscriptions in one particular way. It has an extra filter for batch size. If you choose 0, then the subscription will deliver each event as it occurs. Otherwise, if you pick 10, for example, you'll receive results in batches of 10, but no response until there are 10 results.

Example that would provide an immediate result from the event bus query:

```
subscription {
  busEvents(types: [TimeUpdate], batchSize: 0) {
    event {
      ... on TimeUpdate {
        timestamp
      }
    }
  }
}
```

```

{
  "data": {
    "busEvents": [
      {
        "event": {
          "timestamp": "2023-03-05T15:05:33.177819749Z"
        }
      }
    ]
  }
}
```

## GraphQL playground
For a more approachable interface, GraphQL playground provides an interactive environment for you to build up queries with full integrated documentation. 

For most users this will be the best way to get to grips with the schema. 

Try visiting the GraphQL endpoint of a server. For example, you can try the [GraphQL Playground](https://api.n07.testnet.vega.xyz/graphql/) for Fairground, the Vega operated testnet.

> 

Reviewer: Data Node warns that GraphQL playground is not recommended for production environments:

```
2023-06-13T14:23:39.631+0100    WARN    datanode.gateway.gql    graphql/server.go:206   graphql playground enabled, this is not a recommended setting for production
```

It might be worth noting that this may not be available on mainnet, and recommend testing either on fairground or user running their own data-node. I'm not sure what the best policy would be.

:::

## Queries

Here is a simple query to fetch the current block height and Vega time from a data node:

> ```json

Reviewer 2: **nitpick (non-blocking)**:  should be `graphql` rather than `json` as it's invalid JSON

Reviewer 2: ```suggestion
```graphql
```

No other combinations will work with the pagination, for example if you provide both the `first` and `last` properties, only the `first` will be processed and the `last` will be ignored.

By providing the `first` and `after` properties in the pagination object, you can traverse forward, i.e. start to end, through the paged results. Providing the `last` and `before` properties in the pagination object, you can traverse backward, i.e. end to start, through them.

### Paginated result ordering

By default, results will be returned in newest to oldest order, i.e you will always see the newest data at the top of your results. Therefore if you request the first 10 records, you will receive the 10 newest records, while requesting the last 10 records will return the 10 oldest records.

### Paginated results

Most queries for Vega data could return a lot of results. To help navigate this, Vega’s GraphQL schema implements [cursor based pagination ↗](https://graphql.org/learn/pagination/#pagination-and-edges), which gives the user a unique string that can be passed back to the server to fetch data relative to the last request.

> Before getting into cursor passing, let’s look at a paginated result. You can guess a result will be paginated if the query name ends with `Connection`.

Reviewer 2: **praise (non-blocking)**: I like this description 👍

> ```json

Reviewer 2: **nitpick (non-blocking)**: Same as above

Reviewer 2: ```suggestion
```graphql
```

After fetching this result, it's clear that there are more proposals. You can use `endCursor` from the first result to fetch the next one, which will return the next 1 proposal, and a new cursor:

> ```json

Reviewer 2: **nitpick (non-blocking)**:  You know what's coming...

Reviewer 2: ```suggestion
```graphql
```

```

## Subscriptions

Subscriptions are like queries that keep returning data as it updates. These work similarly to the Observe endpoints in the REST service.

For example, the following query would show a result only when a new proposal came in:

> ```json

Reviewer 2: **nitpick (non-blocking)**:  ...yep...

Reviewer 2: ```suggestion
```graphql
```

A more useful query might be the following one, which receives an update every time the mid-price of the specified market updates:

> ```json

Reviewer 2: **nitpick (non-blocking)**:  ...but nearly...

Reviewer 2: ```suggestion
```graphql
```

subscription {
  marketsData(marketIds: "2dca7baa5f7269b08d053668bca03f97f72e9a162327eebd941c54f1f9fb8f80") {
    midPrice
  }
}
```

### Bus event subscription

The bus event subscription is different from other subscriptions in one particular way. It has an extra filter for batch size. If you choose 0, then the subscription will deliver each event as it occurs. Otherwise, if you pick 10, for example, you'll receive results in batches of 10, but no response until there are 10 results.

Example that would provide an immediate result from the event bus query:

> ```json

Reviewer 2: **nitpick (non-blocking)**:  ...done now.

Reviewer 2: ```suggestion
```graphql
```
